Add clarification for cloud events (#3417)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/pubsub/pubsub-cloudevents.md

Dapr uses CloudEvents to provide additional context to the event payload, enabling features like:

- Tracing
- Content-type for proper deserialization of event data
- Verification of sender application

## CloudEvents example

> A publish operation to Dapr will result in a cloud event envelope containing the following fields:

Reviewer: ```suggestion
A publish operation to Dapr results in a cloud event envelope containing the following fields:
```

## Publish your own CloudEvent

If you want to use your own CloudEvent, make sure to specify the [`datacontenttype`]({{< ref "pubsub-overview.md#setting-message-content-types" >}}) as `application/cloudevents+json`.

> If the CloudEvent that was authored by the app does not contain the minimum required CloudEvent specification, the message will be rejected. If the CloudEvent is valid, Dapr will send it as is and will not add additional fields to it.

Reviewer 2: It would be very helpful to document what that minimum required CloudEvent spec is :)

Reviewer 2: It's also my understanding that Dapr will augment with additional properties if they haven't been specified. Which properties are they?

<img width="601" alt="image" src="https://github.com/dapr/docs/assets/4224880/7e05e602-6ae1-4b9d-9e4d-db4d8ffb9182">

Author: That's correct, updated with the fields data.

Reviewer 2: @yaron2 awesome!

I notice that `datacontenttype` is optional according to the CE spec, as is `dataschema` `subject` and `time`.

If I were to provide any of `dataschema` `subject` `time`, would dapr pass these through or would it strip out?

Equally, If I were to append properties to the envelope that were completely outside of the CE spec, would these get stripped or preserved?

Author: If you pass any field, part of the CE spec or not, Dapr would pass these through.

Reviewer 2: Got it. Might be worth including that in the docs too?

Second thought that _might_ cause confusion for SDK users where `CloudEvent<T>` may not have an ability to carry arbitrary fields. Hmm.

Author: Added a note. Users using the SDK should be mindful of its behavior - from this documentation perspective, the payload is JSON.

## Publish your own CloudEvent

If you want to use your own CloudEvent, make sure to specify the [`datacontenttype`]({{< ref "pubsub-overview.md#setting-message-content-types" >}}) as `application/cloudevents+json`.

> If the CloudEvent that was authored by the app does not contain the [minimum required fields](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#required-attributes) in the CloudEvent specification, the message will be rejected. Dapr will add the following fields to the CloudEvent if they are missing:

Reviewer: ```suggestion
If the CloudEvent that was authored by the app does not contain the [minimum required fields](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#required-attributes) in the CloudEvent specification, the message is rejected. Dapr adds the following fields to the CloudEvent if they are missing:
```

{{< /tabs >}}

## Event deduplication

> When using cloud events created by Dapr, the envelope contains an `id` field which can be used by the app to perform message deduplication. Dapr does not handle deduplication automatically. Dapr supports using message brokers that natively enable message deduplication.

Reviewer: When is says "Dapr supports using message brokers that natively enable message deduplication" should this should say that deduplication can only be done by the broker? Of the 20 brokers Dapr is integrated with, they do not all have deduplication (e.g. In Memory https://docs.dapr.io/reference/components-reference/supported-pubsub/setup-inmemory/)

Author: No, because deduplication can also be done manually via the app. In effect, the text here says that:

1. An app can perform deduplication
2. A message broker can perform deduplication if supported
3. Dapr does not handle deduplication
